Implement percentage step sizes for fans

bdraco opened the pull request:
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

This was deferred from the original implementation of https://github.com/home-assistant/architecture/issues/127 to contain scope (https://github.com/home-assistant/frontend/pull/8216#issuecomment-766355828) cc @Jc2k 

This is a followup PR to address the request to include
step sizes that was original made in related frontend #45407 

The step size is automatically calculated by having the integration provide
the `speed_count` property.  This avoids the need for the integration deal
with `float` values since fans with 7 speeds have a percentage step size of
`14.285714285714286`...

Frontend: https://github.com/home-assistant/frontend/pull/8393
Dev docs: https://github.com/home-assistant/developers.home-assistant/pull/804
User docs: https://github.com/home-assistant/home-assistant.io/pull/16558 and https://github.com/home-assistant/home-assistant.io/pull/16562

Adds a `fan.increase_speed` and `fan.decrease_speed` service to handle
the case where remotes have an up/down button where it would
be difficult to pre-select a percentage

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/16558

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

bdraco commented:
Still need to increase coverage and update demo fans

bdraco commented:
I'm not happy with the floats.  Need to come up with a way to avoid that complexity

bdraco commented:
I think it's better for them to tell the model how many speeds the fan has and it deals with the floats.  Abstracts it away

bdraco commented on homeassistant/components/bond/fan.py:
```diff
@@ -85,6 +85,11 @@ def percentage(self) -> Optional[str]:
+    @property
+    def speed_count(self) -> Optional[int]:
+        """Return the number of speeds the fan supports."""
+        return self._speed_range[1] - self._speed_range[0] + 1
```
This is a bit WET even though its simple. I should make a util

crazyfx1 commented on homeassistant/components/knx/fan.py:
```diff
@@ -68,6 +69,11 @@ def percentage(self) -> Optional[int]:
+    @property
+    def speed_count(self) -> Optional[int]:
+        """Return the number of speeds the fan supports."""
+        return int_states_in_range(self._step_range)
```
Does `int_states_in_range` handle `None` gracefully? `self._step_range` can be None, if the KNX fan supports setting percentage directly.

bdraco replied:
No, it will blow up. The guard should be in the integration though and not the util

bdraco replied:
I added a guard to return the maximum allowed steps in this case (100)

bdraco commented on homeassistant/components/knx/fan.py:
```diff
@@ -68,6 +69,11 @@ def percentage(self) -> Optional[int]:
+    @property
+    def speed_count(self) -> Optional[int]:
+        """Return the number of speeds the fan supports."""
+        return int_states_in_range(self._step_range)
```
```suggestion
        if not self._step_range:
            return None
        return int_states_in_range(self._step_range)
```

bdraco commented on homeassistant/components/knx/fan.py:
```diff
@@ -68,6 +69,13 @@ def percentage(self) -> Optional[int]:
+    def speed_count(self) -> Optional[int]:
+        """Return the number of speeds the fan supports."""
+        if not self._step_range:
+            return None
```
```suggestion
            return 100
```

bdraco commented on homeassistant/components/knx/fan.py:
```diff
@@ -68,6 +69,13 @@ def percentage(self) -> Optional[int]:
+    @property
+    def speed_count(self) -> Optional[int]:
+        """Return the number of speeds the fan supports."""
+        if not self._step_range:
```
```suggestion
        if self._step_range is None:
```

bdraco commented on homeassistant/components/tuya/fan.py:
```diff
@@ -102,6 +102,13 @@ def oscillate(self, oscillating) -> None:
+    def speed_count(self) -> Optional[int]:
+        """Return the number of speeds the fan supports."""
+        if self.speeds is None:
+            return None
```
```suggestion
            return 100
```

bdraco commented on homeassistant/components/knx/fan.py:
```diff
@@ -68,6 +69,13 @@ def percentage(self) -> Optional[int]:
+    def speed_count(self) -> Optional[int]:
+        """Return the number of speeds the fan supports."""
+        if self._step_range is None:
+            return 100
```
```suggestion
            return super().speed_count
```

bdraco commented on homeassistant/components/tuya/fan.py:
```diff
@@ -102,6 +102,13 @@ def oscillate(self, oscillating) -> None:
+    def speed_count(self) -> Optional[int]:
+        """Return the number of speeds the fan supports."""
+        if self.speeds is None:
+            return 100
```
```suggestion
            return super().speed_count
```

bdraco commented:
Will need to come back around and adjust `esphome` after this https://github.com/home-assistant/core/pull/46712

bdraco commented:
Need to rebase for zha

bdraco committed:
Implement percentage step sizes for fans

This is a followup PR to address the request to include
step sizes that was original made in related frontend #45407
https://github.com/home-assistant/frontend/pull/8216#issuecomment-766355828

Adds a fan.increase_speed and fan.decrease_speed to handle
the case where remotes have an up/down button where it would
be difficult to pre-select a percentage

balloob commented on homeassistant/components/fan/services.yaml:
```diff
@@ -100,3 +100,37 @@ set_direction:
+    entity_id:
+      description: Name(s) of the entities to increase speed
+      example: "fan.living_room"
+    percentage_step:
```
Mark as optional.

balloob replied:
```suggestion
    percentage_step:
      required: false
```

balloob replied:
Should we mark it advanced too ?

bdraco replied:
Yes

balloob commented on homeassistant/components/fan/services.yaml:
```diff
@@ -100,3 +100,37 @@ set_direction:
+    entity_id:
+      description: Name(s) of the entities to decrease speed
+      example: "fan.living_room"
+    percentage_step:
```
Mark as optional

balloob replied:
```suggestion
    percentage_step:
      required: false
```

bdraco committed:
Update homeassistant/components/fan/services.yaml

Co-authored-by: Paulus Schoutsen <paulus@home-assistant.io>

bdraco committed:
Update homeassistant/components/fan/services.yaml

Co-authored-by: Paulus Schoutsen <paulus@home-assistant.io>

bdraco commented on homeassistant/components/fan/services.yaml:
```diff
@@ -100,3 +100,39 @@ set_direction:
+      description: Name(s) of the entities to increase speed
+      example: "fan.living_room"
+    percentage_step:
+      required: false
```
```suggestion
      advanced: true
      required: false
```

bdraco committed:
Update homeassistant/components/fan/services.yaml

bdraco commented on homeassistant/components/fan/services.yaml:
```diff
@@ -100,3 +100,39 @@ set_direction:
+      description: Name(s) of the entities to decrease speed
+      example: "fan.living_room"
+    percentage_step:
+      required: false
```
```suggestion
      advanced: true
      required: false
```

bdraco committed:
Update homeassistant/components/fan/services.yaml

bdraco commented:
Thanks!

MartinHjelmare commented on homeassistant/components/fan/__init__.py:
```diff
@@ -246,6 +272,33 @@ async def async_set_percentage(self, percentage: int) -> None:
         else:
             await self.async_set_speed(self.percentage_to_speed(percentage))
 
+    async def async_increase_speed(self, percentage_step=None) -> None:
```
```py
async def async_increase_speed(self, percentage_step: Optional[int] = None) -> None:
```

MartinHjelmare commented on homeassistant/components/fan/__init__.py:
```diff
@@ -246,6 +272,33 @@ async def async_set_percentage(self, percentage: int) -> None:
+        """Decrease the speed of the fan."""
+        await self._async_adjust_speed(-1, percentage_step)
+
+    async def _async_adjust_speed(self, modifier, percentage_step) -> None:
```
```py
async def _async_adjust_speed(self, modifier: int, percentage_step: Optional[int]) -> None:
```

MartinHjelmare commented on homeassistant/components/fan/__init__.py:
```diff
@@ -408,6 +461,19 @@ def percentage(self) -> Optional[int]:
         return 0
 
+    @property
+    def speed_count(self) -> Optional[int]:
```
```py
def speed_count(self) -> int:
```

MartinHjelmare commented on homeassistant/components/fan/__init__.py:
```diff
@@ -408,6 +461,19 @@ def percentage(self) -> Optional[int]:
+        return 100
+
+    @property
+    def percentage_step(self) -> Optional[float]:
```
```py
def percentage_step(self) -> float:
```

MartinHjelmare reviewed (commented):
The `speed_count` property will always return an integer.

bdraco commented:
Typing fixes in https://github.com/home-assistant/core/pull/46790
